src/main.c

```diff
@@ -73,6 +73,85 @@ static void help() {
             "Report bugs to: https://github.com/kspalaiologos/bzip3\n");
 }
 
+static void xwrite(const void * data, size_t size, size_t len, FILE * des) {
+    if (fwrite(data, size, len, des) != len) {
+        fprintf(stderr, "Write error: %s\n", strerror(errno));
+        exit(1);
+    }
+}
+
+static int read_errcheck(void * data, size_t size, size_t len, FILE * des, size_t * out) {
+    *out = fread(data, size, len, des);
+    if (ferror(des)) {
+        fprintf(stderr, "Read error: %s\n", strerror(errno));
+        return -1;
+    }
+    return 0;
+}
+
+/* 0: expected eof
+ * 1: read succeeded
+ * -1: unexpected eof or error
+ */
+static int read_eofcheck(void * data, size_t size, size_t len, FILE * des) {
```

Comment:
it would be cleaner to return the amount of bytes read, or -1 on unexpected EOF/error & use the `read_errcheck` function you had just defined.